feat(forms): Unified Control State Change Events

JeanMeche opened the pull request:
This commit adds a global observable to subscribe to track changes around any `AbstractControl` (and its children).

Fixes #10887

AndrewKushnir commented on goldens/public-api/forms/index.md:
```diff
@@ -29,6 +35,8 @@ export abstract class AbstractControl<TValue = any, TRawValue extends TValue = T
     clearAsyncValidators(): void;
     clearValidators(): void;
+    // (undocumented)
+    readonly controlStateChanges: Observable<ControlEvent<TValue>>;
```
For consistency with Router's observable, I'd propose to consider calling this field `events`:

https://github.com/angular/angular/blob/78e69117f07e8ae11aab3abe0343df815b3649be/packages/router/src/router.ts#L127-L130

JeanMeche commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -991,6 +994,300 @@ describe('reactive forms integration tests', () => {
+    });
+
+    it('should not emit twice the same value', () => {
+      const fc = new FormControl<string|null>('foo', Validators.required);
```
I'm a bit on the fence for this test case. 
* More logic to not emit twice the same value
* But not consistant view other event that have a `emitEvent` option. 

Thoughts @dylhunn

dylhunn replied:
IMO: I think this is an opportunity to improve the behavior, and we should do it

dylhunn reviewed (approved):
I continue to be a bit scared of changing the value/validity code, because it has so many subtle impacts. This looks pretty much as described in the design doc, but let's get it to our GDEs early so they can test it out.

dylhunn commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -7,7 +7,7 @@
 
 import {EventEmitter, ɵRuntimeError as RuntimeError, ɵWritable as Writable} from '@angular/core';
-import {Observable} from 'rxjs';
+import {filter, map, Observable, Subject} from 'rxjs';
```
Can we import `filter` and `map` from `rxjs/operators`? g3 needs this.

dylhunn commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -593,7 +655,9 @@ export abstract class AbstractControl<TValue = any, TRawValue extends TValue = T
    */
-  public readonly valueChanges!: Observable<TValue>;
+  public readonly valueChanges: Observable<TValue> = this.events.pipe(
+      filter((e): e is ValueChangeEvent<TValue> => e instanceof ValueChangeEvent),
```
g3 lint: `e` needs an explicit type or `any`

dylhunn commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -593,7 +655,9 @@ export abstract class AbstractControl<TValue = any, TRawValue extends TValue = T
-  public readonly valueChanges!: Observable<TValue>;
+  public readonly valueChanges: Observable<TValue> = this.events.pipe(
+      filter((e): e is ValueChangeEvent<TValue> => e instanceof ValueChangeEvent),
+      map(e => e.value));
```
g3 lint: `e` needs an explicit type or `any`

dylhunn commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -603,7 +667,9 @@ export abstract class AbstractControl<TValue = any, TRawValue extends TValue = T
    */
-  public readonly statusChanges!: Observable<FormControlStatus>;
+  public readonly statusChanges: Observable<FormControlStatus> = this.events.pipe(
+      filter((e): e is StatusChangeEvent => e instanceof StatusChangeEvent), map(e => e.status));
```
as above: g3 lint: `e` needs an explicit type or `any`

dylhunn reviewed (approved):
A few more nits actually.

reviewed-for: fw-forms, public-api

AndrewKushnir commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -1336,20 +1319,182 @@ export abstract class AbstractControl<TValue = any, TRawValue extends TValue = T
+      this._parent._markAsDirty(opts, changedControl);
+    }
+
+    if (changed) {
```
I'm wondering if we should extend this logic for new components to also check for `opts.emitEvent`, similar to what we have for existing value and status change events and corresponding APIs? Previously APIs like `markAsDirty` didn't emit events, so the `opts.emitEvent` was not relevant, but now they do and for consistency it might be great ti support that. WDYT? // cc @dylhunn

AndrewKushnir commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -1336,20 +1319,182 @@ export abstract class AbstractControl<TValue = any, TRawValue extends TValue = T
+  _updatePristine(opts: {onlySelf?: boolean}, changedControl: AbstractControl): void {
+    const newPristine = !this._anyControlsDirty();
+    const changed = this.pristine !== newPristine;
     (this as Writable<this>).pristine = !this._anyControlsDirty();
```
Looks like we calculate `this._anyControlsDirty()` twice. I'd propose refactoring the code to do it once and reuse the result.

AndrewKushnir commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -1336,20 +1319,182 @@ export abstract class AbstractControl<TValue = any, TRawValue extends TValue = T
+    }
+
+    if (changed) {
+      (this.events as Subject<ControlEvent<TValue>>)
+          .next(new PristineChangeEvent(this.pristine, changedControl));
```
This code is a bit repetitive. I think we can create a protected `emit` function on the `AbstractControl` class and use it in other places like this:
```suggestion
      this.emit(new PristineChangeEvent(this.pristine, changedControl));
```

AndrewKushnir commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -796,20 +875,16 @@ export abstract class AbstractControl<TValue = any, TRawValue extends TValue = T
    * Marks the control and all its descendant controls as `touched`.
    * @see {@link markAsTouched()}
    */
   markAllAsTouched(): void {
```
This function likely needs an `emitEvent` flag too (introduced via `opts` object, similar to other APIs).

AndrewKushnir reviewed (commented):
@JeanMeche thanks for addressing the feedback, the PR looks great 👍 Just left one more comment to add the `emitEvent` option to an API.

AndrewKushnir commented:
@dylhunn FYI, there were some changes in the PR (see https://github.com/angular/angular/pull/54579#discussion_r1536423546) and I've re-requested your review.

AndrewKushnir commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -582,6 +640,20 @@ export abstract class AbstractControl<TValue = any, TRawValue extends TValue = T
+   * value of a parent control (for example if this FormControl is a part of a FormGroup) is updated
+   * later, so accessing a value of a parent control (using the `value` property) from the callback
+   * of this event might result in getting a value that has not been updated yet. Subscribe to the
+   * `events of the parent control instead.
+   * For other event types, the events are emitted after the parent control has been updated
```
```suggestion
   * `events` of the parent control instead.
   * For other event types, the events are emitted after the parent control has been updated.
```

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

atscott commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -582,6 +640,20 @@ export abstract class AbstractControl<TValue = any, TRawValue extends TValue = T
+   * For other event types, the events are emitted after the parent control has been updated.
+   *
+   */
+  public readonly events: Observable<ControlEvent<TValue>> = new Subject();
```
I would suggest have two properties for this instead or you will run into people mocking these events like we do in the Router. You won't need the cast to call `next` in this case either.
```
  private readonly _events = new Subject<ControlEvent<TValue>>();
  public readonly events = this.events.asObservable();
```

AndrewKushnir replied:
Quick question: wouldn't it still be possible to mock events in this case?

atscott replied:
Not so much. The way this happens in the Router is that people cast the property to a subject (just like we do internally) and call `emit` on it. That doesn't work when you use `asObservable()`

atscott commented on goldens/public-api/forms/index.md:
```diff
@@ -769,6 +781,15 @@ export class PatternValidator extends AbstractValidatorDirective {
+    // (undocumented)
+    pristine: boolean;
+    // (undocumented)
+    sourceControl: AbstractControl;
```
@AndrewKushnir @dylhunn How do you feel about renaming this as just `source`? The `control` suffix feels a bit redundant given the type and just `source` more closely matches the browser event `source`/`target` names

AndrewKushnir replied:
Good point, I like `source` as the name 👍

JeanMeche replied:
Let's do that !

atscott commented on packages/forms/src/model/abstract_model.ts:
```diff
@@ -65,6 +66,63 @@ export const DISABLED = 'DISABLED';
+ * @publicApi
+ */
+export class ValueChangeEvent<T> extends ControlEvent<T> {
+  constructor(public value: T, public source: AbstractControl) {
```
Should these properties be `readonly` as well?

dylhunn commented:
@JeanMeche The TGP unfortunately came back red. I will have a look, but maybe let's also book some time to go over the failures when you're free.

dylhunn commented:
@JeanMeche So after some investigation (thanks for having a look @AndrewKushnir!), it turns out aht people are manually calling `emit`:
```
// Manually trigger stage status change
(someControlArray.statusChanges as EventEmitter<string>).emit(
  someControlArray.status,
);
```
🤯🤯🤯
We may need to temporarily support this pattern, but strongly discourage its use. Then we can remove the API in a separate breaking change. WDYT?

JeanMeche commented:
@dylhunn Sounds good to me, I'll make the breaking change in a separate PR !

JeanMeche committed:
feat(forms): Unified Control State Change Events

This commit introduces a new method to subscribe to on every `AbstractControl` subclass.
It allows to track value, pristine, touched and status changes on a given control.

Fixes #10887

jessicajaniuk commented:
This PR was merged into the repository by commit 1c736dc3b258a502360cda40b3a00c07102ccbf5.
